### PR TITLE
initial implementation for configureServer

### LINKED ISSUE #4902: Keystone-next Admin UI refresh loop on nginx reverse proxy
# Bug report

When I start Keystone-next example projects, for example "examples-next/todo", all works well when I open the Admin UI in browser directly (using listening port, `http://example.com:3000/`).
But when I configure reverse proxy via nginx (`https://example.com/`) - I got the infinite loop of page refresh:
```
Something went wrong.
Unhandled Runtime Error
Error: An error occurred when loading Admin Metadata
Call Stack
useKeystone
webpack-internal:///../../../../packages-next/admin-ui/context/dist/admin-ui.esm.js (415:11)
```
And in output of keystone-next console - repeating error:
```
{"level":50,"time":1613990145346,"pid":23190,"hostname":"example.com","name":"graphql","message":"Access denied","locations":[{"line":3,"column":5}],"path":["keystone","adminMeta"],"uid":"cklgg26gj0000w6xxxug1ajz9","name":"GraphQLError","stack":"node_modules/@keystone-next/admin-ui/system/dist/admin-ui.cjs.dev.js:548:19\nrunMicrotasks (<anonymous>)\n"}
```

## To Reproduce

1. Start the  "examples-next/todo" app.
2. Open the Admin UI interface (http://localhost:3000/) directly in browser - all works well, you will see user register form.
3. Configure nginx host with https or even http, and reverse proxy to the app port:
```
location / {
  proxy_pass http://localhost:3000/;
  # Set additional headers on the upstream request
  proxy_set_header   X-Real-IP           $remote_addr;
  proxy_set_header   X-Forwarded-For     $proxy_add_x_forwarded_for;
  proxy_set_header   X-Forwarded-Proto   $scheme;
  proxy_set_header   X-Forwarded-Host    $host;
  proxy_set_header   X-Forwarded-Port    $server_port;
}
```
4. Try to open the proxied domain (https://example.com/) and got the infinite loop.
5. Same problem is with development mode (`yarn dev`).

Full error output:
```
Unhandled Runtime Error

Error: An error occurred when loading Admin Metadata
Call Stack
useKeystone
webpack-internal:///../../../../packages-next/admin-ui/context/dist/admin-ui.esm.js (415:11)
InitPage
webpack-internal:///../../../../packages-next/auth/pages/InitPage/dist/auth.esm.js (273:90)
renderWithHooks
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (14985:27)
mountIndeterminateComponent
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (17811:13)
beginWork
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (19049:16)
callCallback
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (3945:14)
invokeGuardedCallbackDev
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (3994:16)
invokeGuardedCallback
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (4056:31)
beginWork$1
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (23959:28)
performUnitOfWork
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (22771:12)
workLoopSync
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (22702:22)
renderRootSync
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (22665:7)
performSyncWorkOnRoot
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (22288:18)
flushSyncCallbackQueueImpl/<
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (11327:26)
unstable_runWithPriority
webpack-internal:///../../../../node_modules/react-dom/node_modules/scheduler/cjs/scheduler.development.js (646:12)
runWithPriority$1
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (11276:10)
flushSyncCallbackQueueImpl
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (11322:26)
flushSyncCallbackQueue
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (11309:3)
scheduleUpdateOnFiber
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (21888:9)
dispatchAction
webpack-internal:///../../../../node_modules/react-dom/cjs/react-dom.development.js (16139:26)

### PR DESCRIPTION
**Pull request by gautamsi**
closes #4902 

@JedWatson  @timleslie I am not sure if you want to call this function name differently (`configureServer` or `configureExpress` as in v5). initial try at this, hopefully this fixes the issue for trusting the proxy as well as enabling other scenarios like increasing payload size for graphql. I will add changeset if needed also this is very small change, so go ahead and fix this if you want to merge by changing few details.

**This is needed to enable support for custom express server config. I need it to publish new fields based on TinyMCE and CKEditor5. I will have to use the `configureServer` method and as additional step to enable the field but we can use it until you guys figure out a way to let Fields access express instance and configure static path there.**

### CODE REVIEW

**Comment by gautamsi** [comment]
@MurzNN has tested this and found that the value of `trust proxy` is being set but reverted somehow from verbose logs.

```log
keystone-next dev
🤞 Starting Keystone
  express:application set "x-powered-by" to true +0ms
  express:application set "etag" to 'weak' +2ms
  express:application set "etag fn" to [Function: generateETag] +0ms
  express:application set "env" to 'development' +1ms
  express:application set "query parser" to 'extended' +0ms
  express:application set "query parser fn" to [Function: parseExtendedQueryString] +0ms
  express:application set "subdomain offset" to 2 +0ms
  express:application set "trust proxy" to false +0ms
  express:application set "trust proxy fn" to [Function: trustNone] +1ms
  express:application booting in development mode +0ms
  express:application set "view" to [Function: View] +0ms
```

**Review by MurzNN — COMMENTED** [review]
Issue #4902 is solved via other way, so seems 'trust proxy' is not necessary yet, but will be good to have it whatever.

File: examples-next/basic/keystone.ts [hunk 1]
```diff
@@ -52,5 +52,10 @@ export default auth.withAuth(
     //   // store: redisSessionStore({ client: redis.createClient() }),
     //   secret: sessionSecret,
     // }),
+    server: {
+      configureServer(app) {
+        app.set('trust proxy', true);
```
[the remark this answers is not in the document]
Comment (gautamsi) [reply examples-next/basic/keystone.ts:57]: it may only work if you are using this branch

**Comment by JedWatson** [comment]
Hey @gautamsi sorry it's been so long, I wanted to get back to you with some feedback on this PR

I really like the API, and I see the importance of having this implemented. You've also designed it in a way that's consistent with similar config options which is great 👍 

However, not too long ago we technically supported generating Keystone (both the Admin UI **and** the GraphQL API) as a Next.js app that could be deployed entirely to a Serverless environment like Vercel

I'm really keen to get back to that, which this moves us _ever so slightly_ further away from.

Not saying that's a reason to block it, just that how the config options are set up really needs to be thought about comprehensively so they are clear for both use-cases.

The answer is probably that this config option can only coexist with another specifying that the server target is `express` (as opposed to `nextjs` or something similar)

We also need to carefully consider the lifecycle here -- is this configuration for the _just-created_ `app`? or at some point in the middle of the configuration process? that gets super complex, as we've seen in previous keystone versions, and makes me more inclined to sidestep this entirely and make Keystone able to export its own middleware so you can drop it into your own `app` like Next.js does

Per my comment to @timleslie on #4912 I think we need to do some big picture config design, so we can confidently continue adding these features without accidentally conflicting with other planned changes; sorry that's delaying this, it's high on our list to get done but we're also putting most of our focus right now into documenting and cleaning up what's there as a starting point.

Until then this is on the radar but unlikely to get merged, until we've really thought it through and are confident we're not painting ourselves into a corner.

**Comment by JedWatson** [comment]
Further to that previous comment, the other thing to consider (given the lifecycle of setting up an express app is so delicate) is that maybe we should be exposing more server config for things like `trust proxy`

Again, that's not a straight-forward "yes" for me, just wanted to highlight that we really need to do some big picture config design here around the various use cases.

**Comment by gautamsi** [comment]
@JedWatson thanks for clarifying the point. I had some chat with @timleslie on this. I agree that the new config scheme must be well thought per the future plan. I am also interested in serverless. 

That said, you are also right that this kind of request will be more from community. We can work on creating layers which can have certain aspect, like:

1. Have serverless export which **can not** support some of the feature like middleware and this.
2. Export to middleware which can be plugged in the existing express app or have the `keystone-next` cli run it with middleware config including this kind of configuration support.

For now I am thinking of creating `keystonejs-contrib` cli which would allow options not supported by `keystone-next` cli and later migrate back to official one when ready.

**Comment by gautamsi** [comment]
@JedWatson @timleslie should we use this as experimental feature like https://github.com/keystonejs/keystone/pull/5033#issuecomment-792416453

**Comment by gautamsi** [comment]
closing in favor of #6467

**Pull request closed without merging**